### DISCUSSION TITLE
How appropriate is it to store database client in lifespan (request) state?

### DISCUSSION THREAD

**Discussion by hojimat** [fastapi/fastapi#13029] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [x] I commit to help with one of those options 👆

### Example Code

```python
# ALTERNATIVE 1: LIFESPAN STATE
from contextlib import asynccontextmanager
from fastapi import FastAPI, Request
from motor.motor_asyncio import AsyncIOMotorClient


@asynccontextmanager
async def lifespan(app):
    mongo_client = AsyncIOMotorClient("mongo://")
    mongo_db = mongo_client["mydb"]
    yield {"db": mongo_db}
    mongo_client.close()


app = FastAPI(lifespan=lifespan)


@app.get("/")
async def root(request: Request):
    return request.state.db


# ALTERNATIVE 2: DEPENDENCY INJECTION
from contextlib import asynccontextmanager
from fastapi import FastAPI, Depends
from motor.motor_asyncio import AsyncIOMotorClient, AsyncIOMotorDatabase

mongo_client = None

def init_db():
    global mongo_client
    mongo_client = AsyncIOMotorClient("mongo://")

def get_db():
    return mongo_client["mydb"]

def close_db():
    mongo_client.close()
    


@asynccontextmanager
async def lifespan(app):
    init_db()
    yield
    close_db()


app = FastAPI(lifespan=lifespan)

    
@app.get("/")
async def root(db: Annotated[AsyncIOMotorDatabase, Depends(get_db)]):
    return db
```

### Description

I have FastAPI with MongoDB connected via Motor. I open a MongoDB connection at the beginning of an app and close it at the end using lifespan. 

Since I initialize the client inside lifespan, I cannot use dependency injection to `get_db()` without using global variables. But global variables are highly discouraged in Python.

On the other hand, I cannot help but feel weird to pass database client to Request state and hold it with every request.

So, my question is: how appropriate is it to hold such a sensitive information about database client in requests and tag it along with every request all the time? 

A related question is: does `request.state`mean that all values passed to it are tagged along with each request?

Dependency injection seems more appropriate for this use case but it practically requires global variables (unless I use some ORM which would store a reference to the database for me).

Please help me understand the best practice here!

Thank you!

### Operating System

Linux

### FastAPI Version

0.115.0

### Pydantic Version

2.9.2

### Python Version

3.12.7

**Comment by YuriiMotov** [comment, marked as answer]
I tend to use first approach (with `request.state.db`).

I don't see any problems in having `db` in state for every request. It will only contain reference to that object.
The only thing I dislike in this approach is that we need to cast the type of that variable to make type hints working. But it's also doable

**Comment by yauhen-sobaleu** [comment]
Storing the database client in request.state is not appropriate because it violates the principles of a layered architecture (such as Onion Architecture or Clean Architecture). Your HTTP layer (FastAPI request handlers) should not be directly coupled to the infrastructure layer (database connection). It creates tight coupling and makes the code harder to maintain or swap out storage layers. 

Instead of exposing the database client directly, wrap database operations inside an application service and inject it via FastAPI’s dependency injection.

**app.py** example
```python
import typing as t
from contextlib import asynccontextmanager

from fastapi import FastAPI

from src import domain, infra
from src.config import initialize_config
from src.server.registrar import (
    register_routes,
)


@asynccontextmanager
async def lifespan(app: FastAPI) -> t.AsyncGenerator[None, None]:
    # -----------------------------------------------------------------------------------
    # Get config for the specific environment
    # -----------------------------------------------------------------------------------

    config = initialize_config()


    # -----------------------------------------------------------------------------------
    # Define repositories
    # -----------------------------------------------------------------------------------

    conf = config.infra.aws.s3

    s3_repository = infra.S3Repository(
        bucket=conf.bucket,
        region=config.infra.aws.region,
        endpoint_url=config.infra.aws.endpoint,
    )

    # -----------------------------------------------------------------------------------
    # Define application services
    # -----------------------------------------------------------------------------------

    client_service = domain.ClientService(clients_repo=s3_repository)

    # -----------------------------------------------------------------------------------
    # Register services in a service registry
    # -----------------------------------------------------------------------------------

    domain.service_registry.client_service = client_service

    # -----------------------------------------------------------------------------------
    # Start web server
    # -----------------------------------------------------------------------------------

    yield

app = FastAPI(
    title="Your app",
    lifespan=lifespan,
)

register_routes(app)
```

**registry.py** example
```python
from dataclasses import dataclass

from src.domain.client.service import ClientService


class ServiceIsNotRegisteredException(Exception):
    """
    Thrown when a service has not been registered in a service registry
    """

    def __init__(self, service_name: str):
        super().__init__(
            f"Service ({service_name}) is not registered in a service registry"
        )


@dataclass
class ServiceRegistry:
    client_service: ClientService | None = None

service_registry = ServiceRegistry()
```

**dependencies.py** example 
```python
import typing as t

from fastapi import Form
from pydantic import AnyHttpUrl

from src import domain
from src.domain.client.service import ClientService
from src.domain.registry import ServiceIsNotRegisteredException, service_registry


def client_service() -> ClientService:
    """
    Returns instantiated client service
    """
    if (service := service_registry.client_service) is not None:
        return service

    raise ServiceIsNotRegisteredException("client service")
```

**clients.py** example
```python
# define common server dependency in a separate type alias to avoid long signatures
ClientServiceDep = t.Annotated[domain.ClientService, Depends(client_service)]

@router.post(
    "/add",
    summary="Registers a new client",
    status_code=status.HTTP_201_CREATED,
    responses={
        201: {
            "model": AddClientResponse,
            "description": "Client was created successfully",
        },
    },
)
async def add_client(
    service: ClientServiceDep,
    ... params
) -> AddClientResponse:
    await service.add_client(params)
    ...
```

**Comment by alv2017** [comment]
@yauhen-sobaleu, @YuriiMotov 

What happens to the database in case of the ALTERNATIVE 1? When a client calls the API's endpoint? 
(This is something that I do not completely understand)
